samples/sample-07/src/test/java/com/example/sample07/NewConsumerRebalanceProtocolTest.java

```diff
@@ -0,0 +1,325 @@
+
+		rawConsumer.close();
+
+		assertThat(rawConsumerRevokedCount.await(10, TimeUnit.SECONDS)).isTrue();
+		assertThat(config.partitionAssignedLatch.await(10, TimeUnit.SECONDS)).isTrue();
+
+		config.listenerLatch = new CountDownLatch(2);
+		this.template.send(TOPIC_NAME, 0, null, "my-data");
+		this.template.send(TOPIC_NAME, 1, null,"my-data");
+		assertThat(config.listenerLatch.await(20, TimeUnit.SECONDS)).isTrue();
+	}
+
+	KafkaConsumer<Object, String> createRawConsumer() {
+		final Map<String, Object> props = getConsumerProperties();
+		return new KafkaConsumer<>(props);
+	}
+
+	static Map<String, Object> getConsumerProperties() {
+		final Map<String, Object> props = new HashMap<>();
+
+		props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, BytesDeserializer.class.getName());
+		props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+		props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, BROKER);
+		props.put(ConsumerConfig.GROUP_ID_CONFIG, GROUP_ID);
+		props.put(ConsumerConfig.GROUP_PROTOCOL_CONFIG, "consumer");
+		props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+		return props;
+	}
+
+	@EnableKafka
+	@Configuration
+	public static class TestConfig {
+
+		CountDownLatch partitionRevokedLatch = new CountDownLatch(0);
+		CountDownLatch partitionAssignedLatch = new CountDownLatch(1);
+		CountDownLatch listenerLatch = new CountDownLatch(2);
+
+		@KafkaListener(id = GROUP_ID, topics = TOPIC_NAME, autoStartup = "false")
+		void listen(ConsumerRecord<Integer, String> ignored) {
+			listenerLatch.countDown();
+		}
+
+		@Bean
+		KafkaTemplate<Integer, String> template(ProducerFactory<Integer, String> producerFactory) {
+			return new KafkaTemplate<>(producerFactory);
+		}
+
+		@Bean
+		ProducerFactory<Integer, String> producerFactory() {
+			return new DefaultKafkaProducerFactory<>(KafkaTestUtils.producerProps(BROKER));
+		}
+
+		@Bean
+		ConcurrentKafkaListenerContainerFactory<Integer, String>
+		kafkaListenerContainerFactory(ConsumerFactory<Integer, String> consumerFactory) {
+
+			final ConcurrentKafkaListenerContainerFactory<Integer, String> factory =
+					new ConcurrentKafkaListenerContainerFactory<>();
+
+			factory.setConsumerFactory(consumerFactory);
+
+			factory.getContainerProperties().setConsumerRebalanceListener(new ConsumerAwareRebalanceListener() {
+				@Override
+				public void onPartitionsAssigned(Consumer<?, ?> consumer,
+												 Collection<TopicPartition> partitions) {
+					partitionAssignedLatch.countDown();
+				}
+
+				@Override
+				public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+					partitionRevokedLatch.countDown();
+				}
+			});
+			return factory;
+		}
+
+		@Bean
+		ConsumerFactory<Integer, String> consumerFactory() {
+			final Map<String, Object> props = getConsumerProperties();
+			return new DefaultKafkaConsumerFactory<>(props);
+		}
+
+		@Bean
+		public ThreadPoolTaskExecutor threadPoolTaskExecutor() {
```

Comment:
We don't need beans from here and above since they have to be auto-configured by Spring Boot.
When you switch to the `@ServiceConnection` on the container bean definition, everything will be picked up by auto-configuration.